File: benchmarks/comparative/implementations/python/attitude.py
"""
Python (Brahe) attitude conversion benchmarks.
"""

import numpy as np

import brahe

from benchmarks.comparative.implementations.python.base import (
    ensure_eop,
    time_iterations,
)
from benchmarks.comparative.results import TaskResult


def quaternion_to_rotation_matrix(params: dict, iterations: int) -> TaskResult:
    """Benchmark quaternion to rotation matrix conversion using brahe."""
    ensure_eop()
    quaternions = params["quaternions"]

    def run():
        results = []
        for q in quaternions:
            quat = brahe.Quaternion(q[0], q[1], q[2], q[3])
            rm = quat.to_rotation_matrix()
            mat = rm.to_matrix()
            # Flatten 3x3 to row-major list
            results.append([mat[r][c] for r in range(3) for c in range(3)])
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="attitude.quaternion_to_rotation_matrix",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )


def rotation_matrix_to_quaternion(params: dict, iterations: int) -> TaskResult:
    """Benchmark rotation matrix to quaternion conversion using brahe."""
    ensure_eop()
    matrices = params["matrices"]

    def run():
        results = []
        for mat in matrices:
            np_mat = np.array(mat)
            rm = brahe.RotationMatrix.from_matrix(np_mat)
            q = rm.to_quaternion()
            v = q.to_vector(True)  # scalar_first=True -> [w, x, y, z]
            results.append(v.tolist())
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="attitude.rotation_matrix_to_quaternion",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )


def quaternion_to_euler_angle(params: dict, iterations: int) -> TaskResult:
    """Benchmark quaternion to Euler angle (ZYX) conversion using brahe."""
    ensure_eop()
    quaternions = params["quaternions"]

    def run():
        results = []
        for q in quaternions:
            quat = brahe.Quaternion(q[0], q[1], q[2], q[3])
            ea = quat.to_euler_angle(brahe.EulerAngleOrder.ZYX)
            results.append([ea.phi, ea.theta, ea.psi])
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="attitude.quaternion_to_euler_angle",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )


def euler_angle_to_quaternion(params: dict, iterations: int) -> TaskResult:
    """Benchmark Euler angle (ZYX) to quaternion conversion using brahe."""
    ensure_eop()
    angles = params["angles"]

    def run():
        results = []
        for phi, theta, psi in angles:
            ea = brahe.EulerAngle(
                brahe.EulerAngleOrder.ZYX,
                phi,
                theta,
                psi,
                brahe.AngleFormat.RADIANS,
            )
            q = ea.to_quaternion()
            v = q.to_vector(True)  # [w, x, y, z]
            results.append(v.tolist())
        return results

    times, results = time_iterations(run, iterations)

    return TaskResult(
        task_name="attitude.euler_angle_to_quaternion",
        language="python",
        library="brahe",
        iterations=iterations,
        times_seconds=times,
        results=results,
        metadata={
            "library": "brahe",
            "language": "python",
            "version": getattr(brahe, "__version__", "unknown"),
        },
    )

File: benchmarks/comparative/implementations/rust/src/attitude.rs
use brahe::AngleFormat;
use brahe::attitude::{EulerAngle, EulerAngleOrder, Quaternion, RotationMatrix, ToAttitude};
use brahe::math::SMatrix3;
use std::time::Instant;

pub fn quaternion_to_rotation_matrix(
    params: &serde_json::Value,
    iterations: usize,
) -> (Vec<f64>, serde_json::Value) {
    let quaternions: Vec<Vec<f64>> = serde_json::from_value(params["quaternions"].clone()).unwrap();

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(quaternions.len());

        for q in &quaternions {
            let quat = Quaternion::new(q[0], q[1], q[2], q[3]);
            let rm: RotationMatrix = quat.to_rotation_matrix();
            let mat = rm.to_matrix();
            // Row-major flattening
            let mut flat = Vec::with_capacity(9);
            for r in 0..3 {
                for c in 0..3 {
                    flat.push(mat[(r, c)]);
                }
            }
            results.push(flat);
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}

pub fn rotation_matrix_to_quaternion(
    params: &serde_json::Value,
    iterations: usize,
) -> (Vec<f64>, serde_json::Value) {
    let matrices: Vec<Vec<Vec<f64>>> = serde_json::from_value(params["matrices"].clone()).unwrap();

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(matrices.len());

        for mat in &matrices {
            let m = SMatrix3::new(
                mat[0][0], mat[0][1], mat[0][2], mat[1][0], mat[1][1], mat[1][2], mat[2][0],
                mat[2][1], mat[2][2],
            );
            let rm = RotationMatrix::from_matrix(m).unwrap();
            let q: Quaternion = rm.to_quaternion();
            let v = q.to_vector(true); // [w, x, y, z]
            results.push(vec![v[0], v[1], v[2], v[3]]);
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}

pub fn quaternion_to_euler_angle(
    params: &serde_json::Value,
    iterations: usize,
) -> (Vec<f64>, serde_json::Value) {
    let quaternions: Vec<Vec<f64>> = serde_json::from_value(params["quaternions"].clone()).unwrap();

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(quaternions.len());

        for q in &quaternions {
            let quat = Quaternion::new(q[0], q[1], q[2], q[3]);
            let ea: EulerAngle = quat.to_euler_angle(EulerAngleOrder::ZYX);
            results.push(vec![ea.phi, ea.theta, ea.psi]);
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}

pub fn euler_angle_to_quaternion(
    params: &serde_json::Value,
    iterations: usize,
) -> (Vec<f64>, serde_json::Value) {
    let angles: Vec<Vec<f64>> = serde_json::from_value(params["angles"].clone()).unwrap();

    let mut all_times = Vec::with_capacity(iterations);
    let mut first_results = Vec::new();

    for iter in 0..iterations {
        let start = Instant::now();
        let mut results = Vec::with_capacity(angles.len());

        for a in &angles {
            let ea = EulerAngle::new(EulerAngleOrder::ZYX, a[0], a[1], a[2], AngleFormat::Radians);
            let q: Quaternion = ea.to_quaternion();
            let v = q.to_vector(true); // [w, x, y, z]
            results.push(vec![v[0], v[1], v[2], v[3]]);
        }

        all_times.push(start.elapsed().as_secs_f64());

        if iter == 0 {
            first_results = results;
        }
    }

    (all_times, serde_json::to_value(first_results).unwrap())
}
